pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,339 @@
+	DomainSuffix string
+}
+
+func (o *DiscoveryOptions) mixerEnabled() bool {
+	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
+}
+
+// MCPDiscovery provides storage for NotReadyEndpoints
+type MCPDiscovery struct {
+	// [ip:port]config
+	notReadyEndpoints   map[string]*model.Config
+	notReadyEndpointsMu sync.RWMutex
+	options             *DiscoveryOptions
+	networkForRegistry  string
+	ranger              cidranger.Ranger
+}
+
+// NewMCPDiscovery provides a new instance of Discovery
+func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		options:           options,
+	}
+}
+
+// InitNetworkLookup will read the mesh networks configuration from the environment
+// and initialize CIDR rangers for an efficient network lookup when needed
+func (d *MCPDiscovery) InitNetworkLookup(meshNetworks *meshconfig.MeshNetworks) {
+	if meshNetworks == nil || len(meshNetworks.Networks) == 0 {
+		return
+	}
+
+	d.ranger = cidranger.NewPCTrieRanger()
+
+	for n, v := range meshNetworks.Networks {
+		for _, ep := range v.Endpoints {
+			if ep.GetFromCidr() != "" {
+				_, network, err := net.ParseCIDR(ep.GetFromCidr())
+				if err != nil {
+					log.Warnf("unable to parse CIDR %q for network %s", ep.GetFromCidr(), n)
+					continue
+				}
+				rangerEntry := namedRangerEntry{
+					name:    n,
+					network: *network,
+				}
+				_ = d.ranger.Insert(rangerEntry)
+			}
+			if ep.GetFromRegistry() != "" && ep.GetFromRegistry() == d.options.ClusterID {
+				d.networkForRegistry = n
+			}
+		}
+	}
+}
+
+func (d *MCPDiscovery) endpointNetwork(endpointIP string) string {
+	// If networkForRegistry is set then all endpoints discovered by this registry
+	// belong to the configured network so simply return it
+	if len(d.networkForRegistry) != 0 {
+		return d.networkForRegistry
+	}
+
+	// Try to determine the network by checking whether the endpoint IP belongs
+	// to any of the configure networks' CIDR ranges
+	if d.ranger == nil {
+		return ""
+	}
+	entries, err := d.ranger.ContainingNetworks(net.ParseIP(endpointIP))
+	if err != nil {
+		log.Errora(err)
+		return ""
+	}
+	if len(entries) == 0 {
+		return ""
+	}
+	if len(entries) > 1 {
+		log.Warnf("Found multiple networks CIDRs matching the endpoint IP: %s. Using the first match.", endpointIP)
+	}
+
+	return (entries[0].(namedRangerEntry)).name
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
```

Comment:
I'm not entirely sure how the new selection works - @howardjohn  may have more recent experience with this.
But some of the logic for selection is now based only on labels, and quite a few other things changed. We need to look closer at this method - what is still needed and if it is consistent with the current selection in k8s.

Reply:
I think this is fine, we aren't looking up pod info at all here?